File: assets/composition/walking-skeleton/rust-bootstrap/templates/src/domain/greet.rs
//! The core of the greet aggregate. This module is private to
//! `domain` — adapters and deployment units cannot name it; they
//! reach the core only through the contract face's factories and
//! re-exports.

use std::error::Error;
use std::fmt;

/// GreetError reports why a greeting could not be composed.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GreetError {
    /// The command named nobody to address.
    EmptyName,
}

impl fmt::Display for GreetError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GreetError::EmptyName => write!(f, "name must not be empty"),
        }
    }
}

impl Error for GreetError {}

/// Composes the canonical greeting for `name`.
pub fn message(name: &str) -> Result<String, GreetError> {
    let trimmed = name.trim();
    if trimmed.is_empty() {
        return Err(GreetError::EmptyName);
    }
    Ok(format!("Hello, {trimmed}!"))
}

File: assets/composition/walking-skeleton/go-bootstrap-modulith/templates/internal/modules/greeting/internal/domain/internal/greet/greet.go
// Package greet holds the core of the greet aggregate. Everything
// under internal/domain/internal is unimportable outside the domain
// subtree — adapters and deployment units reach it only through the
// factories exported by the domain package.
package greet

import (
	"errors"
	"fmt"
	"strings"
)

// ErrEmptyName reports a greeting with nobody to address.
var ErrEmptyName = errors.New("name must not be empty")

// Message composes the canonical greeting for name.
func Message(name string) (string, error) {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return "", ErrEmptyName
	}
	return fmt.Sprintf("Hello, %s!", trimmed), nil
}
